src/core/net/mdns.cpp

```diff
@@ -3363,15 +3363,22 @@ Error Core::RxMessage::Init(Instance          &aInstance,
         }
     }
 
-    if (mIsUnicast && mIsQuery)
+    if (mIsUnicast)
     {
+        // For queries:
         // Direct Unicast Queries to Port 5353 (RFC 6762 - section 5.5).
         // Responders SHOULD check that the source address in the query
         // packet matches the local subnet for that link and silently ignore
         // the packet if not.
 
-        LogInfo("We do not yet support unicast query to mDNS port");
-        ExitNow(error = kErrorNotCapable);
+        // For responses:
+        // Source Address Check (RFC 6762 - section 11).
+
+        if (!aSenderAddress.GetAddress().IsLinkLocal())
+        {
+            VerifyOrExit(otPlatInfraIfHasOnLinkPrefix(aSenderAddress.GetInfraIfIndex(), &(aSenderAddress.GetAddress())),
```

Comment:
How about using `Get<BorderRouter::InfraIf>().HasAddressWithOnLinkPrefix()`? We can first check that `address infra-if index matches the interface mDNS is running on?)

Reply:
Thanks for feedback @abtink!

For this PR, testing was done by compiling in simulation mode and running the tests.

There is one thing unclear for me at the moment, calling `otPlatMdnsHandleReceive` with `aIsUnicast` flag correctly set, not set `false` by default in `src/posix/platform/mdns_socket.cpp`. Should `recvfrom` be replaced with `recvmsg` or do you know any other/simpler way to get the `destination` ipv6 address, and then determine if the received is a unicast one?

Reply:
As you mentioned, `recvmsg` seems like a good way but it will likely introduce more complexity. 
I don't know of any simpler way (though I must say I'm not super familiar with all POSIX socket functions). :)

Reply:
Hi @abtink , I have found some time to update this PR.
This is definitely not a final implementation, as I am also not super familiar with POSIX socket functions, any feedback is appreciated.
I've updated `infra_if.cpp` from 'src/posix/platform' to handle IPv4 addresses as well.
As for testing, I've managed to use `ot-cli-commissioner`. In this way I could test the `legacy unicast response` feature both on IPv4 and IPv6. I've also managed to tweak a little bit the mDNS implementation from that app to test `Direct Unicast Queries to port 5353, RFC6762 section 5.5`.
Testing was done both on `posix daemon` and `simulation`. On `simulation`, I could not properly test the `direct unicast feature`, `otPlatInfraIfHasOnLinkPrefix` returns false due to the hardcoded infra if ipv6 address and the source ipv6 address that ot-cli-commissioner chooses to use, but it works on `posix daemon`.
Next steps will be to test `Source Address Check` for incoming responses, `RFC6762, section 11`, and I was also thinking about founding an mDNS spoofing tool to test against this, as it is also described in mDNS spec.